storage: fix deadlock in coalesced heartbeats

rjnn opened the pull request:
The `Replica.ReportUnreachable()` code path held `r.raftMu`, and was
called by the heartbeat coalescing goroutine. If `r.raftMu` was
unavailable, this held up all heartbeat traffic. Change reporting
unreachable remote replicas to add to a per-Replica slice of
unreachable remotes, and have the actual reporting happen on
`Replica.tick()`. Closes #10206.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10214)

<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxTuEU52J8Kktr3bu1:-KUxTuEVWdCygrIDbu4_:b-8pcl3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L2220)):*
> ```Go
> 	r.unreachablesMu.Lock()
> 	var unreachableRemotes []roachpb.ReplicaID
> 	copy(unreachableRemotes, r.unreachablesMu.unreachableRemotes)
> ```

This is only copying 0 elements into the empty `unreachableRemotes`. I'm pretty sure that's not what you intended. You want something like:

```
unreachableRemotes := r.unreachablesMu.unreachableRemotes
r.unreachablesMu.unreachableRemotes = nil
```

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUxUAKbqjqeLDAXM08j:b4myl7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
> ```Go
> 	r.raftMu.Lock()
> 	defer r.raftMu.Unlock()
> 	r.mu.internalRaftGroup.ReportUnreachable(uint64(remoteReplica))
> ```

An alternative to this change is to take another look at not holding `raftMu` here. I think @bdarnell said in the past that holding `raftMu` shouldn't be necessary here, but we did so during the lock refactoring to minimize the locking changes. Taking a look at the etcd code, it looks like `raft.RawNode.ReportUnreachable` only touches in-memory state. In fact, I think we should be holding `r.mu` instead as that is the mutex which protects the in-memory state of `raft.RawNode`. 

---

*[pkg/storage/store.go, line 3252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxSwvgSLorPHSbo3Xy:-KUxSwvgSLorPHSbo3Xz:bgn5pqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/store.go#L3252)):*
> ```Go
> 		for _, beat := range beats {
> 			replica := s.mu.replicas[beat.RangeID]
> 			replica.addUnreachableRemoteReplica(beat.ToReplicaID)
> ```

You've violated your declared lock ordering here as `Store.mu` is held but you're about to grab `Replica.unreachablesMu`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxTuEU52J8Kktr3bu1:-KUxWY7n27Nx_XeFapQq:b-qsx7u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L2220)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is only copying 0 elements into the empty `unreachableRemotes`. I'm pretty sure that's not what you intended. You want something like:
> 
> ```
> unreachableRemotes := r.unreachablesMu.unreachableRemotes
> r.unreachablesMu.unreachableRemotes = nil
> ```
</details>
Done. Oops.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUxXFIkdYWRdJEC3WZU:b-qkzdy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> An alternative to this change is to take another look at not holding `raftMu` here. I think @bdarnell said in the past that holding `raftMu` shouldn't be necessary here, but we did so during the lock refactoring to minimize the locking changes. Taking a look at the etcd code, it looks like `raft.RawNode.ReportUnreachable` only touches in-memory state. In fact, I think we should be holding `r.mu` instead as that is the mutex which protects the in-memory state of `raft.RawNode`. 
</details>
I still prefer this overall structure as this way the coalesced heartbeats goroutine does not acquire any replica locks (save for the newly created one which only locks a single data structure). I think the real lesson here is that the coalesced heartbeats goroutine is very fragile, and anything that can delay it for more than 100ms is going to cause lots of cascading failures as heartbeats aren't sent. Regardless of whichever lock is held during the `ReportUnreachable` call, it shouldn't be acquired by the coalesced heartbeats goroutine. I'm fine with that being dictated by whatever locks `Tick()` chooses to hold.

---

*[pkg/storage/store.go, line 3252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxSwvgSLorPHSbo3Xy:-KUxWHeIZ6qaU9VvMU-2:b-jia2io) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/store.go#L3252)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You've violated your declared lock ordering here as `Store.mu` is held but you're about to grab `Replica.unreachablesMu`.
</details>
Fixed the lock ordering comment ( it makes my head hurt).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

tamird commented:
This needs a test, right?

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxTuEU52J8Kktr3bu1:-KUxZlwIU36EvybgSveh:bwmqbdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L2220)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Done. Oops.
</details>
Do we actually need the copy, here? We are still holding raftMu during the call to `ReportUnreachable`, why is it important to release `unreachblesMu`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KUx_QMo4z__9k_pAmCy:ba7ux7s)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUxZfB9bN6DAAn8zKYg:bbe51jl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> I still prefer this overall structure as this way the coalesced heartbeats goroutine does not acquire any replica locks (save for the newly created one which only locks a single data structure). I think the real lesson here is that the coalesced heartbeats goroutine is very fragile, and anything that can delay it for more than 100ms is going to cause lots of cascading failures as heartbeats aren't sent. Regardless of whichever lock is held during the `ReportUnreachable` call, it shouldn't be acquired by the coalesced heartbeats goroutine. I'm fine with that being dictated by whatever locks `Tick()` chooses to hold.
</details>
Not sure if I buy this argument. The coalesced heartbeat goroutine acquires `Store.mu` which has been problematic in the past. 

---

*[pkg/storage/replica.go, line 379 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxZvLRuI4VOCE4PSPp:-KUxZvLSYgRoyvxDALD8:bls5zpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e069eb8aec94ca233503aafc460927ba48a3d09/pkg/storage/replica.go#L379)):*
> ```Go
> 
> 	unreachablesMu struct {
> 		syncutil.TimedMutex
> ```

Given how short you hold this mutex, I think it is overkill to use a `TimedMutex` here. 

---

*[pkg/storage/store.go, line 3252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxSwvgSLorPHSbo3Xy:-KUx_SO6bvGZjHTfD1IV:bx2mqtb) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/store.go#L3252)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Fixed the lock ordering comment ( it makes my head hurt).
</details>
I think you're still violating your declared lock ordering as you're holding `Store.coalescedMu` here.  Perhaps `Store.sendQueuedHeartbeats` should release `Store.coalescedMu` before calling `Store.sendQueuedHeartbeatsToNode`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
It's not clear to me how this specific bug can be tested. Do you want to generally test that unreachable replicas are reported correctly?

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxTuEU52J8Kktr3bu1:-KUxdPKbJdfLW6rb3RWZ:b-ti4lrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L2220)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Do we actually need the copy, here? We are still holding raftMu during the call to `ReportUnreachable`, why is it important to release `unreachblesMu`?
</details>
We're setting it to `nil` when we're done sending, so it's not unreasonable to make the copy to expedite releasing the `unreachablesMu` lock. You're probably right that nothing useful will be done in parallel given `raftMu` is being held here.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUxbbAGblbv6Xf0LwuI:b-nc7fap) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Not sure if I buy this argument. The coalesced heartbeat goroutine acquires `Store.mu` which has been problematic in the past. 
</details>
It has, but my understanding is that the intention is for `Store.mu` to not be locked up for long periods of time. `Replica.mu` is harder to guarantee due to the existence of snapshots. Or is this a false distinction?

---

*[pkg/storage/replica.go, line 379 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxZvLRuI4VOCE4PSPp:-KUxb6TnapuMKB2pjVey:b-d6232) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e069eb8aec94ca233503aafc460927ba48a3d09/pkg/storage/replica.go#L379)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Given how short you hold this mutex, I think it is overkill to use a `TimedMutex` here. 
</details>
OK. Changed to a `Mutex`.

---

*[pkg/storage/store.go, line 3252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxSwvgSLorPHSbo3Xy:-KUxcjLWHC79EiAy0RIa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/store.go#L3252)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think you're still violating your declared lock ordering as you're holding `Store.coalescedMu` here.  Perhaps `Store.sendQueuedHeartbeats` should release `Store.coalescedMu` before calling `Store.sendQueuedHeartbeatsToNode`. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KUxfDujNsOMZlNx1KJq:b-3kdaq8)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 380 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxhf7RESjHDZLQz8h7:-KUxhf7SYr3kbuYL2dk9:b-v8at3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L380)):*
> ```Go
> 	unreachablesMu struct {
> 		syncutil.TimedMutex
> 		unreachableRemotes []roachpb.ReplicaID
> ```

`unreachablesMu.unreachableRemotes` stutters. I think this should be `s/unreachableRemotes/remotes/g`.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUxi9kGKom0R13LicNa:b-o933cb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> It has, but my understanding is that the intention is for `Store.mu` to not be locked up for long periods of time. `Replica.mu` is harder to guarantee due to the existence of snapshots. Or is this a false distinction?
</details>
We don't hold `Replica.mu` while applying snapshots (do we?).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Won't be reviewing closely before bedtime, but the commit message seems to miss the point that the deadlock was due to lock ordering violation (`raftMu < coalescedMu` in one goroutine, `coalescedMu < raftMu` in another goroutine). What's described right now is just a slow lock, not deadlock.

tschottdorf commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUxjaZVNT-UMyWkPmLY:b-xgo2mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We don't hold `Replica.mu` while applying snapshots (do we?).
</details>
I'm not fully paged in on the above discussion, but my gut feeling at least for now is that we shouldn't try to acquire anything else under `Store.mu`, and especially not a `Replica.mu`. At least not these days, when ill-behaved Replicas show very long critical sections.

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KUxjOVU1nDvUP8yuYb1:b-deosif) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
> ```Go
> func (r *Replica) addUnreachableRemoteReplica(remoteReplica roachpb.ReplicaID) {
> 	r.unreachablesMu.Lock()
> 	r.unreachablesMu.unreachableRemotes = append(r.unreachablesMu.unreachableRemotes, remoteReplica)
> ```

not a map? This seems like it'll pile up Replicas en masse if you're unlucky.

---

*[pkg/storage/store.go, line 3249 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxkAapgzMgd8v5-BDY:-KUxkAapgzMgd8v5-BDZ:b-plhslh) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/store.go#L3249)):*
> ```Go
> 
> 	if !s.cfg.Transport.SendAsync(chReq) {
> 		s.mu.Lock()
> ```

You could even grab the replicas under the lock and then do the calls outside of the lock. This is roughly how `newStoreReplicaVisitor` works and it does that because it's easy to do make things bad™ when doing too much work with Store.mu held.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxTuEU52J8Kktr3bu1:-KUxlQz2lGHhk02LRe1I:b-8ii6g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L2220)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> We're setting it to `nil` when we're done sending, so it's not unreasonable to make the copy to expedite releasing the `unreachablesMu` lock. You're probably right that nothing useful will be done in parallel given `raftMu` is being held here.
</details>
My point is that you could write:
```
r.unreachablesMu.Lock()
for _, remoteID := range r.unreachablesMu.remotes {
	r.mu.internalRaftGroup.ReportUnreachable(uint64(remoteID))
}
r.unreachablesMu.remotes = r.unreachablesMu.remotes[:0]
r.unreachablesMu.Unlock()
```
To @tschottdorf point below about this structure getting big: not setting the slice to nil reduces memory churn.

Note also that I renamed stuff here since these `remotes` are certainly not `remoteReplica`s, they are IDs.

---

*[pkg/storage/replica.go, line 2676 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxlZxjji5fDJxdCLJD:-KUxlZxjji5fDJxdCLJE:b932fkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/401f543d68f0b12500689bd73f5451f14e08ac13/pkg/storage/replica.go#L2676)):*
> ```Go
> }
> 
> // reportUnreachable reports the remote replica as unreachable to the internal
> ```

this comment is now incorrect

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

For testing, we should make sure we have a test that fails with the deadlock tag without this change applied. (Do we already have such a test? I'm not sure of the status of the tests in this mode)

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KUySpVZ7qFlFXEIGGx3:b-rh5hyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I'm not fully paged in on the above discussion, but my gut feeling at least for now is that we shouldn't try to acquire anything else under `Store.mu`, and especially not a `Replica.mu`. At least not these days, when ill-behaved Replicas show very long critical sections.
</details>
Acquiring `raftMu` minimizes code changes, but it also minimizes subtlety. I'd rather make this change which is clearly safe than make a locking change that requires looking at implementation details of raft. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KUyUwkk4eecd4494yoL:b4seny3)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KV-R_ZJnywenoqK4rxn:bvjr0n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Acquiring `raftMu` minimizes code changes, but it also minimizes subtlety. I'd rather make this change which is clearly safe than make a locking change that requires looking at implementation details of raft. 
</details>
@bdarnell s/minimizes/maximizes/? Or I didn't understand your comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxTuEU52J8Kktr3bu1:-KV001syGcgLsT49M7PA:b-6ez57k) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L2220)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> My point is that you could write:
> ```
> r.unreachablesMu.Lock()
> for _, remoteID := range r.unreachablesMu.remotes {
> 	r.mu.internalRaftGroup.ReportUnreachable(uint64(remoteID))
> }
> r.unreachablesMu.remotes = r.unreachablesMu.remotes[:0]
> r.unreachablesMu.Unlock()
> ```
> To @tschottdorf point below about this structure getting big: not setting the slice to nil reduces memory churn.
> 
> Note also that I renamed stuff here since these `remotes` are certainly not `remoteReplica`s, they are IDs.
</details>
Reusing the `unreachablesMu.remotes` memory means we'll never release that memory if it somehow grows large. I can't see the memory churn here being a big deal. 

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KV0-AdXEDt5d8ZvcYDx:b-8jp2mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> @bdarnell s/minimizes/maximizes/? Or I didn't understand your comment.
</details>
@bdarnell We look at the implementation details of raft all the time. I'm ok with this change, but please confirm that we should have also been holding `Replica.mu` here. Not doing so would allow someone to call `Replica.RaftStatus()` which only grabs `Replica.mu`. 

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KV0-46hw_mTeZ4435QB:b-p0hbpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> not a map? This seems like it'll pile up Replicas en masse if you're unlucky.
</details>
This is cleared out on every tick, though I suppose there are scenarios where we're not processing ticks for some reason (e.g. `raftMu` is being held excessively). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
We do not have a test that fails with the deadlock tag on. I am having trouble writing such a test, and would appreciate any pointers on how to trigger this code path in a test.

Commit message changed.

---

Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/storage/replica.go, line 380 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxhf7RESjHDZLQz8h7:-KUxjYMJTU_8tl_kpGki:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c651180af5a4b9697132365e7307afb3c01b7c/pkg/storage/replica.go#L380)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `unreachablesMu.unreachableRemotes` stutters. I think this should be `s/unreachableRemotes/remotes/g`.
</details>
Done.

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KV1CxV7KqCOEXpNfkHk:b-b5rnjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is cleared out on every tick, though I suppose there are scenarios where we're not processing ticks for some reason (e.g. `raftMu` is being held excessively). 
</details>
Changed to a map.

---

*[pkg/storage/replica.go, line 2676 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxlZxjji5fDJxdCLJD:-KV19oG5JP911k-J5CPJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401f543d68f0b12500689bd73f5451f14e08ac13/pkg/storage/replica.go#L2676)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment is now incorrect
</details>
Done.

---

*[pkg/storage/store.go, line 3249 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxkAapgzMgd8v5-BDY:-KV1BsUSwiGaK_iwQJdi:bw6vk1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/store.go#L3249)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You could even grab the replicas under the lock and then do the calls outside of the lock. This is roughly how `newStoreReplicaVisitor` works and it does that because it's easy to do make things bad™ when doing too much work with Store.mu held.
</details>
That's how it was (at r0), but I used this (cleaner) syntax here under `Store.mu` because the calls are now very lightweight (since they're just adding to the `unreachablesMu.remotes` queue). I don't feel strongly and can revert the change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KV1ekPmW4lHcaqlKT9J:b-nfhhti)*
<!-- Sent from Reviewable.io -->

tamird commented:
@bdarnell we run a nightly deadlock-enabled stress, so if we had a test that failed with the deadlock tag, we would expect to catch it in one day (though no issues have yet been auto filed by this runner, so it could be broken).

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KV1hdWkoNRUzu6iPUFC:b-fm9vk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
> ```Go
> 	r.unreachablesMu.Lock()
> 	remotes := r.unreachablesMu.remotes
> 	r.unreachablesMu.remotes = nil
> ```

How does this not NPE? This passed CI...which seems to imply that this code is literally dead in our tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KV0Qwn8hKzUmLb5KkJk:bdm3dbr)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KV1oSSYWiZXue9uhcgq:brd3j0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> How does this not NPE? This passed CI...which seems to imply that this code is literally dead in our tests.
</details>
Which part do you expect to NPE?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KV1p57T0wpioiyJGfKj:b-ltqya3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Which part do you expect to NPE?
</details>
The later adding of elements to this nil map.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KV1prxT5TuNBHz2Ce6j:bbxqpbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Changed to a map.
</details>
You need to initialize the map here if it is `nil`.

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KV1pnl18Ivh8Raw1GN-:bhpfjua) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> The later adding of elements to this nil map.
</details>
Oh, I didn't notice it was no longer a slice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@arjunravinarayan I think we don't have any tests which invoke the `ReportUnreachable` code path. I'd try creating a `multiTestContext`-based test which replicates a range to 3 nodes, then trips the `RaftTransport`circuit breaker on the leader to one of the followers. 

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KV1pzda-WOfwW87fENr:b-updkf2)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> I'd try creating a multiTestContext-based test which replicates a range to 3 nodes, then trips the RaftTransportcircuit breaker on the leader to one of the followers.

Argh, `multiTestContext` is such a PITA and we really shouldn't write new tests using it. Major plause if you can make it work with a "real" server (though it might be difficult).

petermattis commented:
> Argh, multiTestContext is such a PITA and we really shouldn't write new tests using it. Major plause if you can make it work with a "real" server (though it might be difficult).

I don't think `TestCluster` gives you access to the `RaftTransport`, though I suppose we could expose that.

tamird commented:
TestCluster?

On Wed, Oct 26, 2016 at 5:37 PM, Tobias Schottdorf <notifications@github.com
> wrote:

> I'd try creating a multiTestContext-based test which replicates a range to
> 3 nodes, then trips the RaftTransportcircuit breaker on the leader to one
> of the followers.
>
> Argh, multiTestContext is such a PITA and we really shouldn't write new
> tests using it. Major plause if you can make it work with a "real" server
> (though it might be difficult).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10214#issuecomment-256484684>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPPMYhlVgsHS_o1FmGjAJEdXdnR58ks5q38gRgaJpZM4Kga7W>
> .
>

Commit:
storage: fix deadlock in coalesced heartbeats.

The Replica.ReportUnreachable() code path held r.raftMu, and was
called by the heartbeat coalescing goroutine, which held
r.coalescedMu, a lock ordering violation. If r.raftMu was deadlocked,
this held up all heartbeat traffic. Change reporting unreachable
remote replicas to add to a per-Replica slice of unreachable remotes,
and have the actual reporting happen on Replica.tick(). Closes #10206.

arjunravinarayan commented:
Added a test.

I've settled for a `multiTestContext` version, so the plauses will alas have to wait for another day. I have verified that:
1. The test fails on the old branch with the deadlock tag, with the deadlock successfully identified.
2. The test passes on this fixed branch with the deadlock tag. Stressraced to 1000 runs.

While adding this test, I found that TestRaftHeartbeats was occasionally flaky, so I refactored it. That fix is in a separate PR.

---

Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KV2-9O_LE1oZueiJsAp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You need to initialize the map here if it is `nil`.
</details>
Done.

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KV2-7qhu6Am3oqlBicU:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Oh, I didn't notice it was no longer a slice.
</details>
Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KV6vzSiFJC20r3Cteyu:b5uxuzq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1484 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7jhnR8pvKPK2Cyci3:-KV7jhnSsp6DWuWn5H-5:b-1lu6nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1484)):*
> ```Go
> 
> 	// Shut down a raft transport via the circuit breaker, and trigger the
> 	// reportUnreachable code path.
> ```

`and wait for a few heartbeats in order to trigger the ...`

---

*[pkg/storage/client_raft_test.go, line 1487 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7iyk7rk7mfB_92Ssh:-KV7iyk7rk7mfB_92Ssi:b-icceeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1487)):*
> ```Go
> 	cb := mtc.transports[0].GetCircuitBreaker(roachpb.NodeID(3))
> 	cb.Break()
> 	time.Sleep(5 * mtc.makeStoreConfig(0).RaftTickInterval)
> ```

Where does the `5` come from? If you want to wait for a certain number of ticks to elapse, see how that is done in `TestRangeQuiescence`.

---

*[pkg/storage/client_raft_test.go, line 1504 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7jzLITWRJHaqVVbhx:-KV7jzLJwWTCD60y_DOB:b-1mtalh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1504)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	mtc.waitForValues(key, []int64{3, 3, 2})
> ```

Is this increment and `waitForValues` accomplishing anything?

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KV7ibshQgwrM5VIHtM6:budv2qv) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Done.
</details>
Not sure what was done, but it wasn't what I was suggesting:

```
if r.unreachablesMu.remotes == nil {
  r.unreachablesMu.remotes = make(map[roachpb.ReplicaID]struct{})
}
r.unreachablesMu.remotes[remoteReplica] = struct{}{}
```

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KV7iRxe0UhVgLKsrVV_:bw3gbfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Fixed.
</details>
You can set to `nil` here and only initialize the map when you're adding and it is non-nil. This gets rid of the need to initialize the map in `newReplica`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1475 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlRMLIznHp05Rxgi0:-KVAlRMLIznHp05Rxgi1:b6illf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1475)):*
> ```Go
> 	leaderID := -1
> 	var initialTerm uint64
> 	for i := 0; i < 3; i++ {
> ```

same comment from #10277; `initialTerm` can be initialized after the loop and `3` is a magic number that you should avoid.

---

*[pkg/storage/client_raft_test.go, line 1489 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlmXfcTk-wOsx0Go4:-KVAlmXgfDQvaOWS9rKB:bljf9yw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1489)):*
> ```Go
> 	time.Sleep(5 * mtc.makeStoreConfig(0).RaftTickInterval)
> 
> 	// Ensure that the leadership has not changed, and that the two remaining
> ```

what is the point of all this? it's not testing the behaviour of `ReportUnreachable`, is it?

---

*[pkg/storage/replica.go, line 547 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlCyJqTvf9_eIG_8x:-KVAlCyKTJ0LKi1ALko5:b-run3pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/replica.go#L547)):*
> ```Go
> 	r.mu.outSnapDone = initialOutSnapDone
> 
> 	r.unreachablesMu.Lock()
> ```

why bother locking this? we don't lock `r.mu` in this constructor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 547 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlCyJqTvf9_eIG_8x:-KVAmRC0fmhZFzvrZNa4:bw3ek1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/replica.go#L547)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why bother locking this? we don't lock `r.mu` in this constructor.
</details>
Correct, though we should just leave `r.unreachablesMu.remotes` nil here and create it when necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 1 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 1475 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlRMLIznHp05Rxgi0:-KVDHLSk5xlkb58qdeEm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1475)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> same comment from #10277; `initialTerm` can be initialized after the loop and `3` is a magic number that you should avoid.
</details>
Done.

---

*[pkg/storage/client_raft_test.go, line 1484 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7jhnR8pvKPK2Cyci3:-KVDHdq3eN8dTnetRzXb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1484)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `and wait for a few heartbeats in order to trigger the ...`
</details>
Done.

---

*[pkg/storage/client_raft_test.go, line 1487 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7iyk7rk7mfB_92Ssh:-KVDHiBq5QgAE8ksg4O7:b-qshrrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1487)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Where does the `5` come from? If you want to wait for a certain number of ticks to elapse, see how that is done in `TestRangeQuiescence`.
</details>
I copied `TestRaftHeartbeats`, the old form of which had the `5` hardcoded by Ben, but your way is much better. Done.

---

*[pkg/storage/client_raft_test.go, line 1489 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlmXfcTk-wOsx0Go4:-KVDLlnyy2N0uwmepJZR:b-73j5lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1489)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what is the point of all this? it's not testing the behaviour of `ReportUnreachable`, is it?
</details>
It's testing that if a single transport fails, heartbeats continue to be sent to the other ranges.

---

*[pkg/storage/client_raft_test.go, line 1504 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7jzLITWRJHaqVVbhx:-KVDL2_Tbqepd5VLCzDB:bwq4gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1504)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is this increment and `waitForValues` accomplishing anything?
</details>
My initial thought was to test that the cluster was functioning, since what we saw on the production cluster was that a deadlock in this code path locked up the entire cluster. Upon reflection, I think you are right. All we need to do is ensure that this isn't locking up heartbeat sending, and for that we just need to check that the leadership is stable.

---

*[pkg/storage/replica.go, line 2680 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxjOVU1nDvUP8yuYb0:-KVDNcw-qu3wCNRbUqBu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97329dd272fd26c5c91a2abd79ef0ef4011c8087/pkg/storage/replica.go#L2680)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Not sure what was done, but it wasn't what I was suggesting:
> 
> ```
> if r.unreachablesMu.remotes == nil {
>   r.unreachablesMu.remotes = make(map[roachpb.ReplicaID]struct{})
> }
> r.unreachablesMu.remotes[remoteReplica] = struct{}{}
> ```
</details>
Done.

---

*[pkg/storage/replica.go, line 2208 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV1hdWkoNRUzu6iPUFB:-KVDNXpwAAaJJ_Gg_vk-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b789ed292b9e9d3fdec8db3e594654c977701c0/pkg/storage/replica.go#L2208)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You can set to `nil` here and only initialize the map when you're adding and it is non-nil. This gets rid of the need to initialize the map in `newReplica`.
</details>
Done.

---

*[pkg/storage/replica.go, line 547 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVAlCyJqTvf9_eIG_8x:-KVDNPbX85oCGFZ9Lmck:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/replica.go#L547)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Correct, though we should just leave `r.unreachablesMu.remotes` nil here and create it when necessary.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the slight tweak to the test.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1487 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7iyk7rk7mfB_92Ssh:-KVGIZJSTU5SlSmkNMr8:b-d7qxf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1487)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> I copied `TestRaftHeartbeats`, the old form of which had the `5` hardcoded by Ben, but your way is much better. Done.
</details>
The code I pointed to is:

```
	ticks := mtc.stores[followerIdx].Metrics().RaftTicks.Count
	for targetTicks := ticks() + int64(2*sc.RaftElectionTimeoutTicks); ticks() < targetTicks; {
		time.Sleep(time.Millisecond)
	}
```

This is different than sleeping because it ensures that the ticks have actually passed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-:-KVGJ-pLoin0rSL5ZGJ4:b-cdofch)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1487 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KV7iyk7rk7mfB_92Ssh:-KVL1B_Au2SteOA5NLK0:b-2618k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7760aef09e3aed6962d858b8dd0bdad318d02a2/pkg/storage/client_raft_test.go#L1487)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The code I pointed to is:
> 
> ```
> 	ticks := mtc.stores[followerIdx].Metrics().RaftTicks.Count
> 	for targetTicks := ticks() + int64(2*sc.RaftElectionTimeoutTicks); ticks() < targetTicks; {
> 		time.Sleep(time.Millisecond)
> 	}
> ```
> 
> This is different than sleeping because it ensures that the ticks have actually passed.
</details>
Aha, got it. This is indeed better. I've gone ahead and also updated #10277 as it has a fixed duration sleep as well (from before).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1482 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KVSvE8RuABkYDZfGfwJ:-KVSvE8S0XkHJBDICchJ:bexgbg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/27e863d90ab0660494778f1c35966cc5ddc38e32/pkg/storage/client_raft_test.go#L1482)):*
> ```Go
> 	initialTerm := mtc.stores[leaderIdx].RaftStatus(1).Term
> 	// Choose a follower index that is guaranteed to not be the leader
> 	followerIdx := len(mtc.stores) + 1 - leaderIdx
> ```

If `leaderIdx = 2` then `followerIdx = 3+1-2 = 2`, so this guarantee doesn't hold.

---

*[pkg/storage/replica.go, line 2667 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214#-KUxUAKbqjqeLDAXM08i:-KVSwLuVD4tRusxW4lAl:b18kmx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5696d0652927c9faf44d0fa9c64019c56d77a0b7/pkg/storage/replica.go#L2667)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> @bdarnell We look at the implementation details of raft all the time. I'm ok with this change, but please confirm that we should have also been holding `Replica.mu` here. Not doing so would allow someone to call `Replica.RaftStatus()` which only grabs `Replica.mu`. 
</details>
OK, if there's already precedent for us doing stuff on the raftGroup without raftMu then this is fine. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10214)*
<!-- Sent from Reviewable.io -->
